consensus/istanbul/backend/handler.go

```diff
@@ -98,28 +142,77 @@ func (sb *Backend) SetBroadcaster(broadcaster consensus.Broadcaster) {
 	sb.broadcaster = broadcaster
 }
 
-func (sb *Backend) NewChainHead() error {
+// SetP2PServer implements consensus.Handler.SetP2PServer
+func (sb *Backend) SetP2PServer(p2pserver consensus.P2PServer) {
+	sb.p2pserver = p2pserver
+	sb.valEnodeTable.p2pserver = p2pserver
+}
+
+// This function is called by worker.go whenever it gets a newWork event.
+func (sb *Backend) NewWork() error {
 	sb.coreMu.RLock()
 	defer sb.coreMu.RUnlock()
 	if !sb.coreStarted {
 		return istanbul.ErrStoppedEngine
 	}
 
-	// If the last block of the epoch has just been added to the blockchain, then
-	// establish 'validator' type connections to all validators in the upcoming epoch
-	// and disconnect from the ones that are no longer in the val set.
-	currentBlock := sb.currentBlock()
-	if istanbul.IsLastBlockOfEpoch(currentBlock.Number().Uint64(), sb.config.Epoch) {
-		sb.logger.Trace("At end of epoch and going to refresh validator peers", "current block number", currentBlock.Number().Uint64())
-		valset := sb.getValidators(currentBlock.Number().Uint64(), currentBlock.Hash())
-		if _, val := valset.GetByAddress(sb.Address()); val == nil {
-			sb.logger.Info("Validators Election Results: Node OUT ValidatorSet")
+	go sb.istanbulEventMux.Post(istanbul.FinalCommittedEvent{})
+	return nil
+}
+
+// This function is called by all nodes.
+// At the end of each epoch, this function will
+//    1)  Output if it is or isn't an elected validator if it has mining turned on.
+//    2)  Refresh the validator connections if it's a proxy or non proxied validator
+func (sb *Backend) NewChainHead(newBlock *types.Block) {
+	if istanbul.IsLastBlockOfEpoch(newBlock.Number().Uint64(), sb.config.Epoch) {
+		sb.coreMu.RLock()
+		defer sb.coreMu.RUnlock()
+
+		valset := sb.getValidators(newBlock.Number().Uint64(), newBlock.Hash())
+
+		// Output whether this validator was or wasn't elected for the
+		// new epoch's validator set
+		if sb.coreStarted {
+			if _, val := valset.GetByAddress(sb.ValidatorAddress()); val != nil {
```

Comment:
Eventually we probably want to look at the BLS key here as well, but I think it's fine to not do that for now.